File: mobile/lib/core/constants/age_categories.dart
/// Categorías de Edad Bovinas - Hacienda Gamelera
/// 4 categorías exactas (NO MODIFICAR sin autorización de Bruno Brito Macedo)

/// Enum de categorías de edad
enum AgeCategory {
  terneros('terneros', 'Terneros', '<8 meses', 0, 8),
  vaquillonasTorillo(
    'vaquillonas_torillos',
    'Vaquillonas/Torillos',
    '6-18 meses',
    6,
    18,
  ),
  vaquillonasToretes(
    'vaquillonas_toretes',
    'Vaquillonas/Toretes',
    '19-30 meses',
    19,
    30,
  ),
  vacasToros('vacas_toros', 'Vacas/Toros', '>30 meses', 30, null);

  const AgeCategory(
    this.value,
    this.displayName,
    this.rangeDescription,
    this.minMonths,
    this.maxMonths,
  );

  /// Valor para almacenamiento/API
  final String value;
  
  /// Nombre para mostrar en UI
  final String displayName;
  
  /// Descripción del rango
  final String rangeDescription;
  
  /// Edad mínima en meses
  final int minMonths;
  
  /// Edad máxima en meses (null = sin límite)
  final int? maxMonths;

  /// Valida si un string es una categoría válida
  static bool isValid(String category) {
    try {
      fromValue(category);
      return true;
    } catch (_) {
      return false;
    }
  }

  /// Obtiene AgeCategory desde string value
  static AgeCategory fromValue(String value) {
    return AgeCategory.values.firstWhere(
      (category) => category.value == value,
      orElse: () => throw ArgumentError('Categoría de edad inválida: $value'),
    );
  }

  /// Obtiene categoría apropiada según edad en meses
  static AgeCategory fromAgeInMonths(int months) {
    if (months < 8) return AgeCategory.terneros;
    if (months >= 6 && months <= 18) return AgeCategory.vaquillonasTorillo;
    if (months >= 19 && months <= 30) return AgeCategory.vaquillonasToretes;
    return AgeCategory.vacasToros;
  }

  /// Verifica si una edad en meses pertenece a esta categoría
  bool containsAge(int months) {
    if (maxMonths == null) {
      return months >= minMonths;
    }
    return months >= minMonths && months <= maxMonths!;
  }
}

/// Lista de todas las categorías para uso en dropdowns
const List<AgeCategory> allAgeCategories = AgeCategory.values;

File: backend/app/core/constants/age_categories.py
"""
Age Categories Constants
4 categorías de edad EXACTAS para Hacienda Gamelera
NO MODIFICAR sin autorización de Bruno Brito Macedo
"""

from datetime import datetime
from enum import Enum


class AgeCategory(str, Enum):
    """
    4 categorías de edad bovinas para Hacienda Gamelera.

    Definidas por Bruno Brito Macedo según prácticas ganaderas locales.
    Usadas para: clasificación automática, reportes SENASAG, análisis GDP.
    """

    TERNEROS = "terneros"  # <8 meses
    VAQUILLONAS_TORILLOS = "vaquillonas_torillos"  # 6-18 meses
    VAQUILLONAS_TORETES = "vaquillonas_toretes"  # 19-30 meses
    VACAS_TOROS = "vacas_toros"  # >30 meses

    @classmethod
    def from_birth_date(cls, birth_date: datetime) -> "AgeCategory":
        """
        Calcula categoría de edad desde fecha de nacimiento.

        Args:
            birth_date: Fecha de nacimiento del animal

        Returns:
            AgeCategory correspondiente según edad en meses

        Examples:
            >>> AgeCategory.from_birth_date(datetime(2024, 4, 1))  # 6 meses
            AgeCategory.TERNEROS
        """
        now = datetime.now()
        age_months = (now.year - birth_date.year) * 12 + (now.month - birth_date.month)

        if age_months < 8:
            return cls.TERNEROS
        if 6 <= age_months <= 18:
            return cls.VAQUILLONAS_TORILLOS
        if 19 <= age_months <= 30:
            return cls.VAQUILLONAS_TORETES
        return cls.VACAS_TOROS

    @classmethod
    def from_age_months(cls, age_months: int) -> "AgeCategory":
        """
        Calcula categoría desde edad en meses.

        Args:
            age_months: Edad del animal en meses

        Returns:
            AgeCategory correspondiente
        """
        if age_months < 8:
            return cls.TERNEROS
        if 6 <= age_months <= 18:
            return cls.VAQUILLONAS_TORILLOS
        if 19 <= age_months <= 30:
            return cls.VAQUILLONAS_TORETES
        return cls.VACAS_TOROS

    @property
    def display_name(self) -> str:
        """Nombre para mostrar en UI."""
        return AGE_CATEGORY_DISPLAY_NAMES[self]

    @property
    def range_description(self) -> str:
        """Descripción del rango de edad."""
        return AGE_CATEGORY_RANGES[self][2]


# Mapeos de categorías
AGE_CATEGORY_DISPLAY_NAMES: dict[AgeCategory, str] = {
    AgeCategory.TERNEROS: "Terneros",
    AgeCategory.VAQUILLONAS_TORILLOS: "Vaquillonas/Torillos",
    AgeCategory.VAQUILLONAS_TORETES: "Vaquillonas/Toretes",
    AgeCategory.VACAS_TOROS: "Vacas/Toros",
}

# Rangos: (min_months, max_months, description)
AGE_CATEGORY_RANGES: dict[AgeCategory, tuple[int, int | None, str]] = {
    AgeCategory.TERNEROS: (0, 8, "<8 meses"),
    AgeCategory.VAQUILLONAS_TORILLOS: (6, 18, "6-18 meses"),
    AgeCategory.VAQUILLONAS_TORETES: (19, 30, "19-30 meses"),
    AgeCategory.VACAS_TOROS: (30, None, ">30 meses"),
}
